### PR TITLE
[core] use fallocate for fallback allocation to avoid SIGBUS

### LINKED ISSUE #16540: [Core] Plasma SIGBUS on shuffling data loader workload
On a shuffling data loader workload that should succeed (with object spilling), workers fail with SIGBUS errors until the job eventually fails when attempting to deserialize an error object.

### Reproduction

1. Clone the `ray_shuffling_data_loader` repository: `git clone git@github.com:ray-project/ray_shuffling_data_loader.git`
2. Checkout the `feat/benchmarks` branch: `git checkout feat/benchmarks`
3. Go to the benchmarks directory: `cd benchmarks`
4. Modify the `file_mounts` in the cluster config to match your local setup.
4. Deploy the cluster: `ray up -y cluster.yaml`
5. Run the following benchmark: `ray exec cluster.yaml "python ~/benchmarks/benchmark.py --num-rows $(( 4 * 10**8 )) --num-files 50 --num-row-groups-per-file 5 --num-reducers 48 --num-trainers 16 --batch-size 250000 --cluster --num-trials 1 --num-epochs 10 --max-concurrent-epochs 2 --data-dir s3://shuffling-data-loader-benchmarks/data/ --stats-dir s3://shuffling-data-loader-benchmarks/stats/ --unique-stats"`

Before any batches are consumed for the first epoch, you should see some "worker died" log entries in your stdout:
```
2021-06-18 18:16:20,400 WARNING worker.py:1121 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. Task ID: 20d3f3cc944a6233ffffffffffffffffffffffff01000000 Worker ID: be83b929e6ff60eeccf28813d11a31db46057bca77f3b3fa88be6ea5 Node ID: feb8aecacc34109681f89304aae86051ce00c1ea00b4e64778223f05 Worker IP address: 172.31.1.17 Worker port: 10021 Worker PID: 104716
```

If you ssh onto the head node, you should find SIGBUS errors in the core worker logs:
```
$ ray attach /path/to/repo/ray_shuffling_data_loader/benchmarks/cluster.yaml
$ cd /tmp/ray/session_latest/logs
$ cat python-core* | grep " E "
[2021-06-18 18:16:17,582 E 106350 19837] logging.cc:440: *** Aborted at 1624040177 (unix time) try "date -d @1624040177" if you are using GNU date ***
[2021-06-18 18:16:18,024 E 106350 19837] logging.cc:440: PC: @                0x0 (unknown)
[2021-06-18 18:16:18,065 E 106350 19837] logging.cc:440: *** SIGBUS (@0x7f2e9e0ec000) received by PID 106350 (TID 0x7f30d0bfd700) from PID 18446744072066351104; stack trace: ***
[2021-06-18 18:16:18,104 E 106350 19837] logging.cc:440:     @     0x7f3d52876e4f google::(anonymous namespace)::FailureSignalHandler()
[2021-06-18 18:16:18,110 E 106350 19837] logging.cc:440:     @     0x7f3d544c6980 (unknown)
[2021-06-18 18:16:18,114 E 106350 19837] logging.cc:440:     @     0x7f3d54251d1b (unknown)
[2021-06-18 18:16:18,120 E 106350 19837] logging.cc:440:     @     0x7f3d52bb4540 execute_native_thread_routine
[2021-06-18 18:16:18,136 E 106350 19837] logging.cc:440:     @     0x7f3d544bb6db start_thread
[2021-06-18 18:16:18,142 E 106350 19837] logging.cc:440:     @     0x7f3d541e471f clone
```

### PR DESCRIPTION
**Pull request by scv119**
Recently we had a SIGBUS bug that fallback allocation crashes with SIGBUS error when /tmp is full. Ideally we'd like to throw some OOM error instead of SIGBUS. 

To address the problem, this PR uses [fallocate](https://man7.org/linux/man-pages/man2/fallocate.2.html) which guarantees that the follow up write access won't fail if fallocate call succeed. Note that this only works for linux. 

Another note is there is a [posix_fallocate](https://man7.org/linux/man-pages/man3/posix_fallocate.3.html) which falls back to emulation on non-linux systems. This is also worth considering but it may ends up with different performance behavior when fallocate/emulation is used. There is some additional caveat:

>        In the glibc implementation, posix_fallocate() is implemented
>        using the fallocate(2) system call, which is MT-safe.  If the
>        underlying filesystem does not support fallocate(2), then the
>        operation is emulated with the following caveats:
> 
>        * The emulation is inefficient.
> 
>        * There is a race condition where concurrent writes from another
>          thread or process could be overwritten with null bytes.
> 
>        * There is a race condition where concurrent file size increases
>          by another thread or process could result in a file whose size
>          is smaller than expected.
> 
>        * If fd has been opened with the O_APPEND or O_WRONLY flags, the
>          function fails with the error EBADF.

Test with shuffler:
```
(pid=36628) Epoch 0 done on consumer 1.
(pid=36628) Starting epoch 1 on consumer 1.
2021-07-06 03:37:57,350	ERROR worker.py:79 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): ray::Consumer.consume() (pid=36616, ip=172.31.10.69)
  File "python/ray/_raylet.pyx", line 493, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 514, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 384, in ray._raylet.raise_if_dependency_failed
ray.exceptions.RayTaskError: ray::shuffle_reduce() (pid=38802, ip=172.31.10.69)
  File "python/ray/_raylet.pyx", line 563, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 564, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 1670, in ray._raylet.CoreWorker.store_task_outputs
  File "python/ray/_raylet.pyx", line 152, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object 31ffddd68e69d833ffffffffffffffffffffffff0100000001000000 in object store because it is full. Object size is 1466584036 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
(pid=36616) Epoch 0 done on consumer 3.
(pid=36616) Starting epoch 1 on consumer 3.
2021-07-06 03:38:02,257	ERROR worker.py:79 -- Unhandled error (suppress with RAY_IGNORE_UNHANDLED_ERRORS=1): ray::Consumer.consume() (pid=36625, ip=172.31.10.69)
  File "python/ray/_raylet.pyx", line 493, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 514, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 384, in ray._raylet.raise_if_dependency_failed
ray.exceptions.RayTaskError: ray::shuffle_reduce() (pid=38830, ip=172.31.10.69)
  File "python/ray/_raylet.pyx", line 563, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 564, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 1670, in ray._raylet.CoreWorker.store_task_outputs
  File "python/ray/_raylet.pyx", line 152, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object 0fb939667ee64409ffffffffffffffffffffffff0100000001000000 in object store because it is full. Object size is 1466027204 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
(pid=36625) Consuming batch on consumer 2 for epoch 0.
```

in raylet.out log:
```
[2021-07-06 03:37:33,758 I 36462 36479] dlmalloc.cc:122: create_and_mmap_buffer(1466028040, /tmp/ray/plasmaXXXXXX)
[2021-07-06 03:37:33,758 D 36462 36479] dlmalloc.cc:158: Preallocating fallback allocation using fallocate
[2021-07-06 03:37:34,540 E 36462 36479] dlmalloc.cc:167: Out of disk space with fallocate error: No space left on device
```

### CODE REVIEW

File: src/ray/object_manager/plasma/dlmalloc.cc [hunk 1]
```diff
@@ -142,6 +142,18 @@ void create_and_mmap_buffer(int64_t size, void **pointer, int *fd) {
     RAY_LOG(INFO) << "Preallocating all plasma memory using MAP_POPULATE.";
     flags |= MAP_POPULATE;
   }
+
+  // For fallback allocation, enable MAP_POPULATE to avoid
+  // SIGBUS issue when runing out of disk space.
+  if (allocated_once && RayConfig::instance().plasma_unlimited()) {
+    if (!MAP_POPULATE) {
+      RAY_LOG(WARNING)
+          << "Fallback allocation: MAP_POPULATE is not available on this platform.";
```
Comment (ericl) [inline src/ray/object_manager/plasma/dlmalloc.cc:151]: This might be too spammy since we will call this very often, consider removing.

**Review by ericl — COMMENTED** [review]
It might be a bit difficult to unit test, but could you at least manually test and check what happens if the disk is full and we trigger a fallback allocation?

**Comment by ericl** [comment]
LGTM pending testing

File: src/ray/object_manager/plasma/dlmalloc.cc [hunk 2]
```diff
@@ -142,6 +142,18 @@ void create_and_mmap_buffer(int64_t size, void **pointer, int *fd) {
...
+    }
+    RAY_LOG(DEBUG) << "Enable MAP_POPULATE for fallback allocation.";
+    flags |= MAP_POPULATE;
+  }
+
   *pointer = mmap(NULL, size, PROT_READ | PROT_WRITE, flags, *fd, 0);
   if (*pointer == MAP_FAILED) {
```
Comment (ericl) [inline src/ray/object_manager/plasma/dlmalloc.cc:158]: Maybe we could change this to say out of disk space?

**Comment by scv119** [comment]
huh SIGBUS still happens with this fix.  investigating.

**Comment by scv119** [comment]
confirmed SIGBUS still happens even if we have `MAP_POPULATE` set:
```
[2021-07-02 03:18:08,731 I 27428 27457] dlmalloc.cc:114: create_and_mmap_buffer(29335560, /tmp/ray/plasmaXXXXXX)
[2021-07-02 03:18:08,731 D 27428 27457] dlmalloc.cc:153: Enable MAP_POPULATE for fallback allocation.
[2021-07-02 03:18:08,745 D 27428 27457] dlmalloc.cc:202: 0x7f4e56891008 = fake_mmap(29335560)
[2021-07-02 03:18:08,745 D 27428 27457] store.cc:373: create object c361e4b8895ba43cffffffffffffffffffffffff0100000023000000 succeeded
```

and later on worker node:

```
[2021-07-02 03:18:25,925 D 45086 45254] gcs_server_address_updater.cc:53: Getting gcs server address from raylet.
[2021-07-02 03:18:26,593 E 45086 96377] logging.cc:440: *** Aborted at 1625195906 (unix time) try "date -d @1625195906" if you are using GNU date ***
[2021-07-02 03:18:26,694 E 45086 96377] logging.cc:440: PC: @                0x0 (unknown)
[2021-07-02 03:18:26,711 E 45086 96377] logging.cc:440: *** SIGBUS (@0x7f444b62e000) received by PID 45086 (TID 0x7f45657fc700) from
```

the proper fix might be installing signal handler...

**Comment by ericl** [comment]
Ah, we must be getting sigbus during the populate then, is that right?
Otherwise I don't see how you can get sigbus if the pages are already
allocated.

**Comment by scv119** [comment]
That's a good point. Actually SIGBUS happens on the worker side, where does MMAP too without the MAP_POPULATE flag. I can try to enable that to see what happens.

**Comment by scv119** [comment]
Doesn't work as expected

**Comment by scv119** [comment]
Using fallocate instead

**Review by ericl — APPROVED** [review]
LGTM. Can we add a test that fills up disk space and checks the error return?

A clever way to do this nondestructively would be to set the fallback dir to /dev/shm as well.

**Comment by clarkzinzow** [comment]
Also @scv119 have you confirmed that this fixes https://github.com/ray-project/ray/issues/16540?

**Comment by scv119** [comment]
@clarkzinzow this turns SIGBUS crash to OOM exceptions, but I'm not sure if shuffle loader could recover from this OOM. I can confirm ray-project/ray_shuffling_data_loader#14 fixed the issue.

**Comment by rkooo567** [comment]
The fix is to use the large disk for spilling. I think this will help you figuring out the problem earlier (but we shouldn't raise OOM, but out of disk space error instead).

**Comment by rkooo567** [comment]
> LGTM pending testing (it would be also great to add the workload to nightly tests, though I'm not sure how easy it would be to replicate the out of disk issue. We might consider an alternate workload that specifically stresses disk space):

We can probably do this by using a smaller instance (and catching the designated exception as a success condition)

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
@ericl @scv119  Btw, how are we deleting the fallback allocated files now? (and is the behavior same after using the fallocat?)

**Comment by scv119** [comment]
yup I think the behavior is the same.

**Comment by ericl** [comment]
Tests seem to be failing on 
```




============================= test session starts ==============================
  | platform linux -- Python 3.6.13, pytest-5.4.3, py-1.10.0, pluggy-0.13.1 -- /opt/miniconda/bin/python3
  | cachedir: .pytest_cache
  | rootdir: /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/python/ray/tests/test_plasma_unlimited.runfiles/com_github_ray_project_ray
  | plugins: asyncio-0.15.1, rerunfailures-10.1, sugar-0.9.4, timeout-1.4.2
  | collecting ... collected 8 items
  |  
  | ::test_fallback_when_spilling_impossible_on_put PASSED                   [ 12%]
  | ::test_spilling_when_possible_on_put PASSED                              [ 25%]
  | ::test_fallback_when_spilling_impossible_on_get PASSED                   [ 37%]
  | ::test_spilling_when_possible_on_get PASSED                              [ 50%]
  | ::test_task_unlimited PASSED                                             [ 62%]
  | ::test_task_unlimited_multiget_args PASSED                               [ 75%]
  | ::test_fd_reuse_no_memory_corruption PASSED                              [ 87%]
  | ::test_fallback_allocation_failure ================================================================================
 

<br class="Apple-interchange-newline">```

**Pull request merged**